### PR TITLE
Feature: implementing the circular buffer library, a data structure.

### LINKED ISSUE #52: Circular Buffer Abstraction
Implement a dynamically allocated circular buffer that can be used by all the other communication protocols.

Specification:
- Dynamically allocated array of pointers
- Stores pointers to structs
- Make sure to do shallow copies

### PR DESCRIPTION
**Pull request by ghost**
Resolves #52

# Circular Buffer

This library defines a new type `CircularBuffer` and declares CircularBuffer-manipulating/inspecting functions:
- `GR_CircularBuffer_Create(uint16_t capacity) -> CircularBuffer*`
- `GR_CircularBuffer_Free(CircularBuffer** buffer) -> void`
- `GR_CircularBuffer_GetCapacity(CircularBuffer* buffer) -> uint16_t`
- `GR_CircularBuffer_GetCurrentSize(CircularBuffer* buffer) -> uint16_t`
- `GR_CircularBuffer_IsFull(CircularBuffer* buffer) -> bool`
- `GR_CircularBuffer_IsEmpty(CircularBuffer *buffer) -> bool`
- `GR_CircularBuffer_Push(CircularBuffer* buffer, void* object) -> uint8_t`
- `GR_CircularBuffer_Pop(CircularBuffer* buffer) -> void*`
- `GR_CircularBuffer_Peek(CircularBuffer* buffer) -> void*`

To use this library with CMake, one needs to link the library `CircularBuffer_Lib` as an `INTERFACE`  to his or her targets with `target_link_libraries()`.

Example usage:
``` c
#include "circularBuffer.h"
#include <stdio.h>

int main(){
	CircularBuffer* buffer_ptr = GR_CircularBuffer_Create(10);

    int arr[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};

	for(int i = 0; i < 10; i++){
		GR_CircularBuffer_Push(buffer_ptr, &arr[i]);
	}
	for(int i = 0; i < 10; i++){
		int* tmp = GR_CircularBuffer_Pop(buffer_ptr);
		printf("%d\n", *tmp);
	}

	GR_CircularBuffer_Free(&buffer_ptr);
}
```
Output:
```stdout
0
1
2
3
4
5
6
7
8
9
```

### CODE REVIEW

File: Lib/Utils/CircularBuffer/Inc/circularBuffer.h
Comment (dchansen06) [inline Lib/Utils/CircularBuffer/Inc/circularBuffer.h]: Please create usable function names, instead of `cbOps.create()` please make the header file have function names more verbose `GR_CircularBuffer_Create()`.

Similarly I do not see a reason to wrap functions in this case in structs, especially since you are not preserving any form of state, if they are clearly named they should be able to be accessed directly from calling functions.

File: Lib/Utils/CircularBuffer/Src/circularBuffer.c
[the remark this answers is not in the document]
Comment (dchansen06) [reply Lib/Utils/CircularBuffer/Src/circularBuffer.c]: It may be nice to have a few functions that return if the circular buffer is empty or not, if it is full or not, and the current (used) size of the circular buffer.

**Comment by XuChe777** [comment]
The new implementation is now complete. The signature of the initializer function changes from `GR_CircularBuffer_Create(uint16_t)` to `GR_CircularBuffer_Create(uint16_t, CircularBufferMode)`. `CircularBufferMode` is a new `enum` type with possible values `CIRCULAR_BUFFER_KEEP_ITEM_ON_OVERWRITE` and `CIRCULAR_BUFFER_FREE_ITEM_ON_OVERWRITE`.

@anthony9975 @Shrav816 You can now update the test codes. Please re-request a review for this pull request once you are finished.

**Comment by dchansen06** [comment]
When you get a chance, can you please make the title of this PR a little more descriptive? Thanks!

**Comment by XuChe777** [comment]
**Final TODO!**: Address sanitizer.

File: G4PERTESTING/Core/Src/main.c [hunk 1]
```diff
@@ -19,6 +19,7 @@
 /* Includes ------------------------------------------------------------------*/
 #include "main.h"
 #include "adc.h"
+#include "circularBuffer.h"
```
Comment (dchansen06) [inline G4PERTESTING/Core/Src/main.c]: Please put our includes inside of `USER CODE BEGIN Includes` / `USER CODE END Includes` to help preserve the HAL/LL boundary

File: Lib/Utils/CircularBuffer/Src/circularBuffer.c [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	// For specifications see the header file
+	CircularBuffer *buffer = malloc(sizeof(CircularBuffer));
+	buffer->head = 0;
+	buffer->tail = 0;
+	buffer->capacity = capacity;
+	buffer->buffer = calloc(capacity, sizeof(void *));
+	return buffer;
+}
+
+void GR_CircularBuffer_Free(CircularBuffer *buffer)
```
Comment (dchansen06) [inline Lib/Utils/CircularBuffer/Src/circularBuffer.c]: Freeing a buffer should ideally set it to null but if that makes life overcomplicated then ignore this comment

File: Lib/Utils/CircularBuffer/Test/ToDo.md
Comment (dchansen06) [inline Lib/Utils/CircularBuffer/Test/ToDo.md]: Please add null tests for passing in null pointers or passing in valid pointers that have the buffer nulled

**Comment by XuChe777** [comment]
> > The failing checks seem to be caused by [a49c2c0](https://github.com/Gaucho-Racing/Firmware/commit/a49c2c08b187ea3c8c5ad5ba31418a53a70d6034) on g4-peripheral-test-setup. Otherwise, all issues here are resolved.
> 
> 
> 
> Yes #48 is critically broken at the moment. The lib looks great to me though, you can do one of two things:
> 
> - Wait for #48 to be fixed and then merge
> 
> - Change to target `main` and then merge (not including any of #48)
> 
> 
> 
> The second option is more correct / preferable but I have no objections if you want to wait either

I will try to switch to target main tomorrow.

**Review by dchansen06 — APPROVED** [review]
Lgtm
